pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go

```diff
@@ -0,0 +1,150 @@
+		p.apiBindingIndexerInitError = err
+		return
+	}
+
+	// just in case the plugin gets init multiple times in case of an error
+	p.apiBindingIndexerInitError = nil
+	p.SetReadyFunc(informers.Apis().V1alpha1().APIBindings().Informer().HasSynced)
+	p.apiBindingIndexer = informers.Apis().V1alpha1().APIBindings().Informer().GetIndexer()
+}
+
+func (p *crdNoOverlappingGVRAdmission) ValidateInitialization() error {
+	if p.apiBindingIndexerInitError != nil {
+		return fmt.Errorf(PluginName+" plugin failed to initialize %q APIBindings indexer, err = %v", byWorkspace, p.apiBindingIndexerInitError)
+	}
+	if p.apiBindingIndexer == nil {
+		return fmt.Errorf(PluginName + " plugin needs an APIBindings indexer")
+	}
+	return nil
+}
+
+// Validate checks if the given CRD's Group and Resource don't overlap with bound CRDs
+func (p *crdNoOverlappingGVRAdmission) Validate(ctx context.Context, a admission.Attributes, _ admission.ObjectInterfaces) error {
+	if a.GetResource().GroupResource() != apiextensions.Resource("customresourcedefinitions") {
+		return nil
+	}
+	if a.GetKind().GroupKind() != apiextensions.Kind("CustomResourceDefinition") {
+		return nil
+	}
+	clusterName, err := request.ClusterNameFrom(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve cluster from context: %w", err)
+	}
+	// ignore CRDs targeting system and non-root workspaces
+	if clusterName == apibinding.ShadowWorkspaceName || clusterName == logicalcluster.New("system:admin") || !clusterName.HasPrefix(tenancyv1alpha1.RootCluster) {
+		return nil
+	}
+
+	crd, ok := a.GetObject().(*apiextensions.CustomResourceDefinition)
+	if !ok {
+		return fmt.Errorf("unexpected type %T", a.GetObject())
+	}
+	apiBindingsForCurrentClusterName, err := p.listAPIBindingsFor(clusterName)
+	if err != nil {
+		return err
+	}
+	var notReadyAPIBindings []string
+	for _, apiBindingForCurrentClusterName := range apiBindingsForCurrentClusterName {
+		if !conditions.IsTrue(apiBindingForCurrentClusterName, apisv1alpha1.BindingUpToDate) {
+			notReadyAPIBindings = append(notReadyAPIBindings, apiBindingForCurrentClusterName.Name)
+			// give it a try anyway, we fail on not ready bindings
+		}
+		for _, boundResource := range apiBindingForCurrentClusterName.Status.BoundResources {
+			if boundResource.Group == crd.Spec.Group && boundResource.Resource == crd.Spec.Names.Plural {
+				return admission.NewForbidden(a, fmt.Errorf("cannot create %q CustomResourceDefinition with %q group and %q resource because it overlaps with a bound CustomResourceDefinition for %q APIBinding in %q logical cluster",
+					crd.Name, crd.Spec.Group, crd.Spec.Names.Plural, apiBindingForCurrentClusterName.Name, clusterName))
+			}
+		}
+	}
+	if len(notReadyAPIBindings) > 0 {
+		return admission.NewForbidden(a, fmt.Errorf("cannot create %q CustomResourceDefinition because the following APIBinding werern't ready and werern't checked for conflicting GVRs = %v", crd.Name, notReadyAPIBindings))
+	}
```

Comment:
still think this whole if clause is not right. Being in !BindingUpToDate is normal, not a reason to stop CRD creation.

Reply:
seems like we need to trade off availability over consistency. My intention was to make sure we have seen all bounded CRDs before admitting a CRD. I guess you prefer the former or is there any other condition we should check? (`!InitialBindingCompleted`?)

Reply:
resources not bound via APIBinding status do not exist :)

Reply:
Same with `!InitialBindingCompleted`. At that point there are no bound resources yet. So no conflict.

There is the second part with the APIBinding controller that it needs similar logic for the inverse check.

Reply:
yeah, at this point it is a best-effort admission